### ISSUE TITLE
Legacy applications: option to enable keeping ResultSet open after transaction closed

### ISSUE THREAD

**Issue by barulheira** [OpenLiberty/open-liberty#12717]
We have a legacy application that reads records in a "paged" fashion. On each new page, more records are retrieved. After each paging, the current transaction may be closed.
When this happens, Open Liberty closes the resultset, and no record reads are allowed anymore.

We need an option to configure Open Liberty (or the data source) to keep resultsets open after the transaction is closed (`connection.close()`), enabling us to retrieve the remaining records.
It's ugly, OK. But it would allow adopting Open Liberty to update such legacy applications.

Such feature works by default in other application servers: JBoss 4, Payara and Tomee.
It doesn't work in Open Liberty and WildFly.

This is a minimal test case to reproduce this behavior:
```
DataSource ds = InitialContext.doLookup("java:comp/env/jdbc/db");
Connection connection = ds.getConnection();
PreparedStatement preparedStatement = connection.prepareStatement("select 1 from dual");
ResultSet resultSet = preparedStatement.executeQuery();
connection.close();
System.out.println(resultSet.next());
```
Success: prints "true"
Failure: throws Exception

**Comment by aguibert** [comment]
hi @barulheira, thanks for raising this issue. I'm very surprised that other app servers allow this behavior, because it is explicitly prohibited by the JDBC spec.

According to JDBC spec, closing a connection must close all associated statements:

> 9.4.4.1 Connection.close
An application calls the method Connection.close() to indicate that it has
finished using a connection. All Statement objects created from a given
Connection object will be closed when the close method for the Connection
object is called.
Once a Connection has been closed, any attempt to access any of its methods with
the exception of the close(), isClosed() or isValid() methods will result in
a SQLException being thrown.

And likewise, closing a statement must close all associated result sets:

> 13.1.4 Closing Statement Objects
Closing a Statement object will close and invalidate any instances of ResultSet
produced by that Statement object. The resources held by the ResultSet object
may not be released until garbage collection runs again, so it is a good practice to
explicitly close ResultSet objects when they are no longer needed.
Once a Statement has been closed, any attempt to access any of its methods with
the exception of the isClosed or close methods will result in a SQLException
being thrown.
These comments about closing Statement objects apply to PreparedStatement
and CallableStatement objects as well.

And lastly, calling any method on a closed ResultSet should result in an error:

> 15.2.5 Once a ResultSet has been closed, any attempt to access any of its methods with
the exception of the isClosed or close methods will result in a SQLException
being thrown. ResultSetMetaData instances that were created by a ResultSet
that has been closed are still accessible.

----

Aside from JDBC spec, it would be dangerous to allow such behavior because depending on how many rows have been fetched from the DB on the initial query, calling `resultSet.next()` could trigger a FETCH from the DB, which would require the connection to be usable, and result in an error if it's not. My guess is that this happens to work in your legacy application because all data is fetched immediately after the query (either because the data set is small, or your fetch size is very large).

You should be able to achieve similar behavior by making some slight changes to your application. Essentially looping over `rs.next()` and eagerly parsing out all required data from the ResultSet immediately after invoking the prepared statement query. Even though you'd be eagerly parsing out data into a separate object, this ought to be no less efficient than what is currently happening, since you'd be getting an error if all of the data wasn't fetched immediately after the query anyway.

**Comment by barulheira** [comment]
Thanks, @aguibert  for your kind remarks. I agree very much with your explanation.

I'd like to present you some real world usage cases we have met (with other application servers):

Although JDBC spec states that closing a statement should close the related resultset, it doesn't actually (we use Oracle JDBC drivers). We have to make sure the resultset is closed before closing the related statement. (I suppose it is a bug in the driver or data source provider... but whatever.)

When using a connection pool, the underlying connection is not actually closed - and for that reason its statements and resultsets aren't actually closed likewise. I suppose it's up to the data source provider to close them properly, either immediately or when it's required to. That's possibly the difference of behavior between those providers.

So, we'll follow your advice, and try to find a way to fetch eagerly the records before closing the transaction. It won't be easy. The framework doesn't know in advance which are the columns that are to be fetched, so we must provide a way for it to remember the colums fetched for the first record. We'll keep you informed on the achievements.

**Comment by njr-11** [comment]
@barulheira there is some inconsistency in the question which needs to be cleared up here.  You asked about keeping a result set open across transaction close but then the code sample goes on to equate the 'closing' of a transaction with Connection.commit, which is not an equivalent thing.  If you are just looking to achieve the behavior of keeping a ResultSet open across transactions, that behavior _is_ defined by the JDBC specification and is possible in OpenLiberty - at least to the extent that the JDBC driver itself supports it.

Here is an example:

```
DataSource ds = InitialContext.doLookup("java:comp/env/jdbc/db");
Connection connection = ds.getConnection();
connection.setHoldability(ResultSet.HOLD_CURSORS_OVER_COMMIT);
connection.setAutoCommit(false);
PreparedStatement preparedStatement = connection.prepareStatement("select 1 from dual");
ResultSet resultSet = preparedStatement.executeQuery();
connection.commit();
System.out.println(resultSet.next());
connection.commit();
connection.close();
```

**Comment by barulheira** [comment]
@njr-11 Yes, it sounds confusing, because the minimal case reproduces what happens in the framework of our legacy application, that is:
 
1. it runs in a JavaEE environment, using the container's connection pool;
2. after each transaction (commit), the framework _closes_ the connection, turning it back to the pool -- although the database connection itself is _not_ closed -- and starts a new transaction retrieving a connection from the pool, which may be the same or not.

Hence, that minimal case _reproduces_ what happens in the framework; it's not how it's coded in the application.

**Comment by Thihup** [comment]
I've managed to get your minimal case to run in Wildfly. I've had to set two properties from the datasource:
```xml
<timeout>
  <use-try-lock>-1</use-try-lock>
</timeout>
<statement>
  <track-statements>false</track-statements>
</statement>
```

So, in Open Liberty as a workaround, you could use the JDBC resource adaptor from Iron Jacamar
```xml
<server>
    <featureManager>
        <feature>jca-1.7</feature>
    </featureManager>

    <library id="IronJacamarLibs">
        <file name="ironjacamar-core-api-1.4.22.Final.jar"/>
        <file name="ironjacamar-jdbc-1.4.22.Final.jar"/>
        <file name="jboss-logging-3.4.1.Final.jar"/>
    </library>

    <resourceAdapter id="IronJacamar" location="jdbc-local-1.4.22.Final.rar">
        <classloader commonLibraryRef="IronJacamarLibs"/>
    </resourceAdapter>

    <connectionFactory jndiName="jdbc/systextil">
        <containerAuthData user="user" password="pass" />
        <properties.IronJacamar>
            <ConnectionURL>url</ConnectionURL>
            <TrackStatements>false</TrackStatements>
            <UseTryLock>-1</UseTryLock>
        </properties.IronJacamar>
    </connectionFactory>
</server>
```
Maybe you need to adjust the libraries. You can find which properties the Iron Jacamar supports in the _META-INF/ra.xml_ file inside the _jdbc-local-1.4.22.Final.rar_ file
